File: pay-later-go/db/report.go
package db

import "database/sql"

// MerchantDiscount computes the total discount for a merchant
func (db *DB) MerchantDiscount(merchant Merchant) (float64, error) {
	var discount sql.NullFloat64
	db.db.Raw(
		`
		SELECT sum(amount) * merchants.discount_percentage / 100 as discount
		FROM transactions
		INNER JOIN merchants on merchants.id = transactions.merchant_id
		WHERE merchant_id = ?
		`,
		merchant.ID,
	).Scan(&discount)
	return discount.Float64, nil
}

// UserDues computes the total dues of a user
func (db *DB) UserDues(user User) float64 {
	var userTotalDues sql.NullFloat64
	db.db.Raw(
		`
		SELECT sum(amount) as total_amount
		FROM transactions
		WHERE user_id = ?
		`,
		user.ID,
	).Scan(&userTotalDues)
	return userTotalDues.Float64
}

// UsersAtCreditLimit lists all users who have reached their credit limit
func (db *DB) UsersAtCreditLimit() []User {
	var users []User
	db.db.Raw(
		`
		SELECT users.*
		FROM transactions
		INNER JOIN users on users.id = transactions.user_id
		GROUP BY user_id
		HAVING sum(amount) = users.credit_limit
		`,
	).Scan(&users)
	return users
}

// UserDue encapsulates each users name and their total due amount
type UserDue struct {
	Name string
	Due  float64
}

// UsersTotalDues lists all users and their total due amount
// Additionally this also includes the total sum of all dues by all users
func (db *DB) UsersTotalDues() []UserDue {
	var userDues []UserDue
	db.db.Raw(
		`
		SELECT users.name, sum(amount) as due
		FROM transactions
		INNER JOIN users on users.id = transactions.user_id
		GROUP BY user_id
		UNION ALL
		SELECT 'total', sum(amount) as due
		FROM transactions
		`,
	).Scan(&userDues)
	return userDues
}

File: pay-later-rs/src/db/report.rs
use anyhow::Result;
use sqlx::{Pool, Sqlite};

pub struct Report {}

impl Report {
    pub async fn merchant_discount(merchant_name: String, pool: &Pool<Sqlite>) -> Result<f64> {
        let discount: (f64,) = sqlx::query_as(
            "SELECT sum(amount) * merchants.discount_percentage / 100 as discount
            FROM transactions
            INNER JOIN merchants on merchants.id = transactions.merchant_id
            WHERE merchant_id = (SELECT id FROM merchants WHERE name = $1)
            ",
        )
            .bind(merchant_name)
            .fetch_one(pool)
            .await?;
        Ok(discount.0)
    }

    pub async fn user_dues(user_name: String, pool: &Pool<Sqlite>) -> Result<f64> {
		let dues: (f64,) = sqlx::query_as(
            "SELECT sum(amount) as total_amount
            FROM transactions
            WHERE user_id = (SELECT id FROM users WHERE name = $1)
            ",
        )
            .bind(user_name)
            .fetch_one(pool)
            .await?;
        Ok(dues.0)
    }

    pub async fn users_at_credit_limit(pool: &Pool<Sqlite>) -> Result<Vec<String>> {
        let users: Vec<(String,)> = sqlx::query_as(
            "SELECT users.name
            FROM transactions
            INNER JOIN users on users.id = transactions.user_id
            GROUP BY user_id
            HAVING sum(amount) = users.credit_limit
            ",
        )
            .fetch_all(pool)
            .await?;
        Ok(users.iter().map(|u| u.clone().0).collect())
    }

    pub async fn users_total_dues(pool: &Pool<Sqlite>) -> Result<Vec<(String, f64)>> {
        let users: Vec<(String, f64)> = sqlx::query_as(
            "SELECT users.name, sum(transactions.amount) as due
            FROM users
            LEFT JOIN transactions on users.id = transactions.user_id
            GROUP BY user_id
            UNION ALL
            SELECT 'total', sum(amount) as due
            FROM transactions
            ",
        )
            .fetch_all(pool)
            .await?;
        Ok(users)
    }
}
